### PR TITLE
Add BooleanType support to pyfunc spark udf

### PR DESCRIPTION
**Pull request by BenWilson2**
Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

## Related Issues/PRs

#xxx

## What changes are proposed in this pull request?

Add support for Spark BooleanType (and, by extension, bool type and np.bool) for spark_udf for pyfunc

## How is this patch tested?

- [X] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add BooleanType (bool) support for pyfunc spark_udf return types.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [X] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [X] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by BenWilson2** [comment]
Running a test to see if BooleanType can be supported

File: tests/pyfunc/test_spark.py [hunk 1]
```diff
@@ -134,6 +142,7 @@ def test_spark_udf(spark, model_path):
         "double": (DoubleType(), np.number),
         "long": (LongType(), int),
         "string": (StringType(), None),
+        "bool": (BooleanType(), bool),
```
Comment (harupy) [inline tests/pyfunc/test_spark.py:145]:
```suggestion
        "boolean": (BooleanType(), bool),
```

File: tests/pyfunc/test_spark.py [hunk 2]
```diff
@@ -147,6 +156,8 @@ def test_spark_udf(spark, model_path):
                 expected = prediction_df.select_dtypes(np_type)
                 if tname == "float":
                     expected = expected.astype(np.float32)
+                if tname == "bool":
```
Comment (harupy) [inline tests/pyfunc/test_spark.py:159]:
```suggestion
                if tname == "boolean":
```

**Comment by harupy** [comment]
Manual testing:

```python
import random

import mlflow
from pyspark.sql import SparkSession
from pyspark.sql.types import BooleanType
from sklearn.datasets import load_iris
from sklearn.datasets import load_iris

spark = SparkSession.builder.getOrCreate()


class BooleanModel:
    def predict(self, X):
        return [bool(random.randint(0, 1))] * len(X)


with mlflow.start_run():
    model_info = mlflow.sklearn.log_model(Model(), "model")

pyfunc_udf = mlflow.pyfunc.spark_udf(spark, model_info.model_uri, result_type=BooleanType())

X, y = load_iris(as_frame=True, return_X_y=True)
df = spark.createDataFrame(X)
df.withColumn("prediction", pyfunc_udf(*df.columns)).show()
```

```
22/09/29 19:28:36 WARN Utils: Your hostname, haru-Z590-S01 resolves to a loopback address: 127.0.1.1; using 192.168.0.177 instead (on interface enp3s0)
22/09/29 19:28:36 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/09/29 19:28:36 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")
2022/09/29 19:28:38 WARNING mlflow.pyfunc: Calling `spark_udf()` with `env_manager="local"` does not recreate the same environment that was used during training, which may lead to errors or inaccurate predictions. We recommend specifying `env_manager="conda"`, which automatically recreates the environment that was used to train the model and performs inference in the recreated environment.
+-----------------+----------------+-----------------+----------------+----------+
|sepal length (cm)|sepal width (cm)|petal length (cm)|petal width (cm)|prediction|
+-----------------+----------------+-----------------+----------------+----------+
|              5.1|             3.5|              1.4|             0.2|      true|
|              4.9|             3.0|              1.4|             0.2|      true|
|              4.7|             3.2|              1.3|             0.2|      true|
|              4.6|             3.1|              1.5|             0.2|      true|
|              5.0|             3.6|              1.4|             0.2|      true|
|              5.4|             3.9|              1.7|             0.4|      true|
|              4.6|             3.4|              1.4|             0.3|      true|
|              5.0|             3.4|              1.5|             0.2|      true|
|              4.4|             2.9|              1.4|             0.2|      true|
|              4.9|             3.1|              1.5|             0.1|     false|
|              5.4|             3.7|              1.5|             0.2|     false|
|              4.8|             3.4|              1.6|             0.2|     false|
|              4.8|             3.0|              1.4|             0.1|     false|
|              4.3|             3.0|              1.1|             0.1|     false|
|              5.8|             4.0|              1.2|             0.2|     false|
|              5.7|             4.4|              1.5|             0.4|     false|
|              5.4|             3.9|              1.3|             0.4|     false|
|              5.1|             3.5|              1.4|             0.3|     false|
|              5.7|             3.8|              1.7|             0.3|     false|
|              5.1|             3.8|              1.5|             0.3|     false|
+-----------------+----------------+-----------------+----------------+----------+
```

**Comment by harupy** [comment]
cc @wenfeiy-db

File: docs/source/models.rst [hunk 3]
```diff
@@ -2074,26 +2074,29 @@ numeric column as a double. You can control what result is returned by supplying
...
+  ``float32`` is returned or an exception is raised if there are no numeric columns.
 * ``'double'`` or DoubleType_: The leftmost numeric result cast to
-  ``double`` is returned or exception is raised if there is no numeric column.
+  ``double`` is returned or an exception is raised if there are no numeric columns.
 * ArrayType_ ( FloatType_ | DoubleType_ ): Return all numeric columns cast to the
-  requested. type. Exception is raised if there are numeric columns.
-* ``'string'`` or StringType_: Result is the leftmost column converted to string.
-* ArrayType_ ( StringType_ ): Return all columns converted to string.
-
-.. _IntegerType: https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.types.IntegerType.html#pyspark.sql.types.IntegerType
```
Comment (BenWilson2) [inline docs/source/models.rst:2091]: These links are all broken

File: mlflow/pyfunc/__init__.py [hunk 4]
```diff
@@ -868,14 +872,16 @@ def spark_udf(spark, model_uri, result_type="double", env_manager="local"):
     should_use_spark_to_broadcast_file = not (is_spark_in_local_mode or should_use_nfs)
     env_root_dir = _get_or_create_env_root_dir(should_use_nfs)
 
+    result_type = "boolean" if result_type == "bool" else result_type
```
Comment (BenWilson2) [inline mlflow/pyfunc/__init__.py:875]: "boolean" is not a python type. Added a resolver here for usability / prevent users from getting annoyed.

Comment (BenWilson2) [reply mlflow/pyfunc/__init__.py:875]: This is also intended to mirror how feature store is handling `score_batch()` with resolving `bool` -> `boolean`.

**Comment by wenfeiy-db** [comment]
Thanks for adding this boolean support! 
Just to confirm, this change is going to be included in next mlflow release (2.0?) and so included in next MLR version. Is this correct?

**Comment by dbczumar** [comment]
> Thanks for adding this boolean support! Just to confirm, this change is going to be included in next mlflow release (2.0?) and so included in next MLR version. Is this correct?

Yep!

**Review by dbczumar — APPROVED** [review]
LGTM!

**Pull request merged**